### PR TITLE
Partially implement flash.display.BitmapData

### PR DESCRIPTION
**Pull request by CUB3D**
Partially implements flash.diplay.BitmapData (#290)

This implements the following methods/properties:
- width
- height
- transparent
- rectangle
- loadBitmap()
- constructor()
- clone()
- getPixel()
- getPixel32()
- setPixel()
- setPixel32()
- colorTransform()
- dispose()
- fillRect()
- floodFill()
- getColorBoundsRect()
- copyChannel()
- MovieClip::attachBitmap()

And stubs the following methods:
- applyFilter
- copyPixels
- draw
- generateFilterRect
- hitTest
- merge
- paletteMap
- perlinNoise
- pixelDissolve
- threshold
- scroll

Currently the backing data is stored as an array of ARGB pre-multiplied alpha pixels to match the behaviour of Flash Player which sometimes causes inaccuracies when using colours with alpha values as a result of converting to and from this format, this should only cause minor visual inaccuracies for most use cases but *might* cause some issues in future (specifically with methods like threshold which work on the pre-multiplied data directly).

This pr also adds support for constructor failure, such as when constructing a BitmapData with invalid bounds e.g. ```var x = new flash.display.BitmapData(3000, 3000);``` which results in the variable getting set to undefined.

### CODE REVIEW

File: core/src/avm1/globals/bitmap_data.rs
Comment (Herschel) [inline core/src/avm1/globals/bitmap_data.rs:140]: `ok()` swallows an error, instead there should be `?` somewhere, here's a pathological example:
```
var o = {valueOf: function() {
	throw "BAD";
	return 3;
}};
trace(bData.getPixel(o, o));
```

File: core/src/avm1/globals/bitmap_data.rs
Comment (Herschel) [inline core/src/avm1/globals/bitmap_data.rs:430]: ditto, need `?` instead of `ok()`

File: core/src/avm1/object/bitmap_data.rs
Comment (Herschel) [inline core/src/avm1/object/bitmap_data.rs:16]: `get_blue` -> `blue` etc.

File: core/src/avm1/globals/bitmap_data.rs
Comment (Herschel) [inline core/src/avm1/globals/bitmap_data.rs:58]: `get_height` => `height`

Let's go with the Rust naming convention of `height` for getters and `set_height` for setters. (I know we haven't been super consistent with this yet, but I'd like to start moving in that direction to keep a consistent style in the codebase).

File: core/src/avm1/globals/bitmap_data.rs
Comment (Herschel) [inline core/src/avm1/globals/bitmap_data.rs:142]: I think `if let (Some(x), Some(y))` is more clear than using zip here.

**Review by Herschel — CHANGES_REQUESTED** [review]
Thanks so much for the work and patience (there's been a lot of PRs to chew threw recently!)

I have to do a more thorough review, but the main issue is that `BitmapData` can be updated live even after they've been attached to the display list, for example:
```
import flash.display.BitmapData;
var bData = new BitmapData(5, 5, true, 0);

createEmptyMovieClip("clip1", 0);
createEmptyMovieClip("clip2", 1);

clip1.attachBitmap(bData, 0);
clip1._xscale = clip1._yscale = 500;
clip1._x = clip1._y = 50;

clip2.attachBitmap(bData, 0);
clip2._xscale = clip2._yscale = 500;
clip2._x = clip2._y = 100;

bData.setPixel32(3, 0, 0xffff0000)
```
[bitmapdata.zip](https://github.com/ruffle-rs/ruffle/files/5611605/bitmapdata.zip)
And both bitmap display object update "live" to show the newly drawn pixel, and this is common way of using BitmapDatas. Currently this PR will show nothing in this case. This involves having some ability to update the textures on the rendering backend. 

Some other nits:
 * Let's avoid the `unwrap`s in the BitmapData methods, because in weirdo cases, people could use `BitmapData.prototype.fn.apply({});` and the underlying object won't be a BitmapData.
 * Let's use `foo` and `set_foo` for getter/setter names, instead of `get_foo` and `set_foo`.
 * I'm not sure about adding the ConstructorFailure case; instead of an hidden "failure" case, I wonder if this can be fixed more generally along with #1469 by allowing built-in constructors that modify the underlying object. (In this case, it would just change it to undefined). But that might be for a separate PR.

File: core/src/avm1/object/bitmap_data.rs
Comment (Herschel) [inline core/src/avm1/object/bitmap_data.rs:156]: This would be better returned as a `&[Color]` and let the caller decide whether to clone it.

File: core/src/avm1/object/bitmap_data.rs
Comment (Herschel) [inline core/src/avm1/object/bitmap_data.rs:152]: Seems bad to make a `Vec` for each pixel just to swizzle the channels.

File: core/src/backend/render.rs
Comment (Herschel) [inline core/src/backend/render.rs:156]: This should probably return a `Result` if it could fail on some backends. This can avoid the `Results` later on.

**Comment by Herschel** [comment]
Thinking about it more, I believe there could be a struct `BitmapData` that exists in the `core` module:
 * [Nearly all of the methods are shared between AS2 and AS3](https://help.adobe.com/en_US/FlashPlatform/reference/actionscript/3/flash/display/BitmapData.html), so these should be exposed as methods on some `BitmapData` struct so that both AVMs can share them.
* The display and AVM objects can hold a `GcCell<BitmapData>` to reference this, allowing the same bitmap data to be shared between multiple bitmap display objects.
* `BitmapData` can own the pixel buffer and a dirty flag. Before/during rendering, a bitmap can check the dirty flag and tell the backend to update the texture if necessary.

**Comment by kmeisthax** [comment]
Having a `BitmapData` struct would also be useful for software video decoders - at the very least, I noticed that there's no way to just funnel bits into the renderer. Going to hold off on workarounds until this PR lands.

**Comment by Herschel** [comment]
Thanks a lot! I made some changes:
 * Implemented Bitmap.smoothing
 * Avoid creating textures in `update_texture`; instead, we can update the previous texture when possible
 * WebGL wasn't working because it expected all bitmaps to have a character ID.
Some more TODO for the future:
 * I think the pixel data should live completely on the core side, i.e. the backends don't really need the bitmap data once it's uploaded to the GPU. This would involve moving more of the bitmap decoding logic to core, and the Bitmap in the library would hold the pixel data.
 * `RenderBackend::get_bitmap_pixels` would then only be used to writeback data from the GPU texture to the bitmap data (this would be needed for BitmapData.draw)
 * I just noticed the backends use character IDs to handle bitmaps in the SWF... this will blow up with multiple SWFs loaded, oops!

Going to go ahead and merge, thanks for your work!

**Pull request merged**
